subprojects/core-api/src/main/java/org/gradle/api/component/ComponentWithCapabilities.java

```diff
@@ -0,0 +1,30 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.gradle.api.component;
+
+import org.gradle.api.Incubating;
+
+import java.util.List;
+
+/**
+ * Represents a component which provides some capabilities.
+ *
+ * @since 4.7
+ */
+@Incubating
+public interface ComponentWithCapabilities extends SoftwareComponent {
```

Comment:
This is kind of odd, because a component can declare capabilities that it doesn't itself provide. `ComponentWithCapabilities` makes it sound like the component _has_ the capability, not just _defines_ a capability.

Also, does it make sense for a particular _variant_ to define a capability, as opposed to the component?
Is it possible to put the capability definitions on `UsageContext` instead?

Reply:
Yeah I see what you mean here. I just followed the pattern that exists in native (lots of `ComponentWith` interfaces). I reckon we also have interfaces that say `HasSomething` and not `WithSomething`, so I think it's not _that_ wrong. It's a component that _has capability declarations_, so maybe I should rename to `WithCapabilityDeclarations`? It's a bit mouthful.